File: course/MA792-002/CppExamples/exception.cpp
// File: exception.cpp
//       runs under g++ version 2.8.1 ("add gcc281")

#include <iostream>
#include <stdexcept>

using namespace std;

// class header info
class Integer
  {int x;
   public:
   Integer() : x(0) {}
   Integer(const int i) : x(i) {}
   inline Integer operator/(const Integer j)
      throw(std::invalid_argument); // indicate what the fun may throw
                         // this is not necessary but helps the caller
   friend ostream& operator<<(ostream& os, const Integer& i);
  };
// implementation
inline Integer Integer::operator/(const Integer j)
      throw(std::invalid_argument)
      {if(j.x == 0)
         throw(std::invalid_argument(
                 "in Integer::operator/: division by zero")
              );
       else return this->x/j.x;
      }
ostream& operator<<(ostream& os, const Integer& i)
  {os << i.x; return os; }
int main(void)
{int x=1, y=0;
 // cout << x/y << endl;
 try
 { Integer three = 3; Integer two = 2; Integer zero = 0;
   cout << three/two << endl;
   cout << three/zero << endl;
 }
 catch (std::invalid_argument& m)
     { cout << "Exception: " << m.what() << endl;
     }
}

File: course/MA792-002/JavaExamples/exception.java
public class exception {

static public class Integer {
// encapsulated to save a file, use Integer name
   private int x;
   public Integer() {x = 0;}
   public Integer(int i) {x = i;}
   public Integer div(Integer j) throws ArithmeticException
         {if(j.x == 0)
            throw new ArithmeticException(
                 "in Integer.div: division by zero");
          else return new Integer(this.x/j.x);
         }
   public String toString()
     {return String.valueOf(x); }
} // end Integer

public static void main(String[] args)
{try
 { Integer three = new Integer(3);
   Integer two = new Integer(2);
   Integer zero = new Integer();
   System.out.println(three.div(two).toString());
   System.out.println(three.div(zero).toString());
 }
 catch (ArithmeticException m)
     { System.out.print("Exception: ");
       System.out.println(m.getMessage());
     }
 finally
     { System.out.println("Continue...");
     }
}// end main
}// end exception
